Library/Homebrew/extend/os/linux/requirements/java_requirement.rb

```diff
@@ -0,0 +1,31 @@
+require "language/java"
+
+class JavaRequirement < Requirement
+  default_formula "jdk"
+
+  env do
+    next unless @java_home
+    env_java
+    java_home = Pathname.new(@java_home)
+    if (java_home/"include").exist? # Oracle JVM
+      env_oracle("linux")
+    end
+  end
+
+  private
+
+  def preferred_java
```

Comment:
@rwhogg To me that implies that `preferred_java` should make use of `satisfies_version` to check each one and then the ordering doesn't actually matter that much; it'll look for the right Java in whatever place it exists. In that case I'd probably go for `JAVA_HOME`, then `java_home_cmd`, then `jdk`, then `which`. You could maybe also abstract both `java_home_cmd` and `jdk` into methods so one or the other is used.